blackbird_python/blackbird/tests/test_utils.py

```diff
@@ -0,0 +1,388 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for the utils module"""
+from textwrap import dedent
+from collections import OrderedDict
+
+import pytest
+
+import numpy as np
+import sympy as sym
+
+from blackbird import loads
+from blackbird.listener import RegRefTransform
+from blackbird.utils import to_DiGraph, match_template, TemplateError
+
+
+class TestDiGraph:
+    """Test conversion of programs to DAGs"""
+
+    def test_no_args(self):
+        """Test case where operations have no arguments"""
+        program = loads(dedent(
+            """\
+            name prog
+            version 1.1
+
+            Vac | 0
+            """
+        ))
+
+        res = to_DiGraph(program)
+        assert len(res) == 1
+
+        nodes = res.nodes().data()
+        assert nodes[0]['name'] == 'Vac'
+        assert nodes[0]['modes'] == (0,)
+        assert nodes[0]['args'] == []
+        assert nodes[0]['kwargs'] == {}
+
+    def test_args(self):
+        """Test case where operations have arguments"""
+        program = loads(dedent(
+            """\
+            name prog
+            version 1.1
+
+            Sgate(0.54) | 1
+            BSgate(0.432, phi=0.54) | [1, 2]
+            """
+        ))
+
+        res = to_DiGraph(program)
+        assert len(res) == 2
+
+        nodes = res.nodes().data()
+        assert nodes[0]['name'] == 'Sgate'
+        assert nodes[0]['modes'] == (1,)
+        assert nodes[0]['args'] == [0.54]
+        assert nodes[0]['kwargs'] == {}
+
+        assert nodes[1]['name'] == 'BSgate'
+        assert nodes[1]['modes'] == (1, 2)
+        assert nodes[1]['args'] == [0.432]
+        assert nodes[1]['kwargs'] == {'phi': 0.54}
+
+        assert list(res.edges()) == [(0, 1)]
+
+    def test_regrefs(self):
+        """Test case where operations have regref arguments"""
+        program = loads(dedent(
+            """\
+            name prog
+            version 1.1
+
+            Dgate(0.54) | 0
+            MeasureX | 0
+            Sgate(2*q0) | 1
+            """
+        ))
+
+        res = to_DiGraph(program)
+        assert len(res) == 3
+        assert list(res.edges()) == [(0, 1), (1, 2)]

```

Comment:
what is the information content of these edges? It seems to just build edges in temporal order...

Reply:
ok I see, the edge (1,2) is there because we are using the result from command 1 measurement as input to command 2.
In that case, it is worthwhile to add in some more tests related to topology, i.e., make sure edges (don't) appear when commands (don't) depend on one another

Reply:
:+1:

Reply:
`test_no_dependence` has been added to `test_utils.py`.